crypto/store/loader_file.c

```diff
@@ -209,32 +213,169 @@ static OSSL_STORE_INFO *try_decode_PKCS12(const char *pem_name,
+                 * checking if the application locale indicates UTF-8 encoded
+                 * input) to BMPString of the pass phrase, and failing that, it
+                 * falls back on the OpenSSL <1.1.0 behaviour.
+                 *
+                 * The correct thing to do for PKCS#12 would be to convert the
+                 * pass phrase from the application's locale to a wide character
+                 * string (and hope this takes care of any normalization), and
+                 * and then simply recode that to BMPString.  That's not what
+                 * OpenSSL's PKCS#12 routines do today, though.
+                 *
+                 * We need to cater for all these possible encodings, so we
+                 * create three variants of the pass phrase:
+                 */
+
+                /*
+                 * Check variant 1: use what we get as is.
+                 * It may be perfectly legitimate UTF-8, which will be decoded
+                 * correctly, and it may be perfectly legitimate ISO-8859-1 or
+                 * ASCII, which will be correctly encded by the <1.1.0 fallback
+                 * in the majority of cases.
+                 */
+                pass = (char *)tpass;
+#ifdef OPENSSL_DEBUG_KEYGEN
+                fprintf(stderr, "Trying password as is:\n");
+                h__dump((unsigned char *)pass, strlen(pass));
+#endif
+                if (PKCS12_verify_mac(p12, pass, strlen(pass)))
+                    goto p12_parse;
+
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199409L
+                /*
+                 * Check variant 2: locale to UTF-8 translation
+                 * Convert the pass phrase from the current application locale
+                 * via wide chars to UTF-8.  This should ensure that we can
+                 * read containers protected with correctly encoded pass
+                 * phrases (i.e. PKCS#12 files from other softwares as well)
+                 */
+                {
+                    wchar_t wpass[PEM_BUFSIZE];
+                    size_t wpass_len;
+                    unsigned char *pass_utf8 = NULL;
+                    size_t utf8_count, i, j;
+
+                    if ((wpass_len = mbstowcs(wpass, (char *)tpass,
+                                              OSSL_NELEM(wpass)))
+                        == (size_t) -1) {
+                        SYSerr(SYS_F_MBSTOWCS, errno);
+                        goto p12_end;
+                    }
+
+                    /*
+                     * We use our own rather than iconv for two reasons:
+                     * 1. We're not sure iconv is available "everywhere"
+                     * 2. We assume that any normalization is done by
+                     *    mbstowcs() and that a naïve conversion to UTF8
+                     *    is therefore harmless.
+                     */
+                    utf8_count = 0;
+                    for (i = 0; i < wpass_len; i++)
```

Comment:
What happens if (despite it not being strictly permitted) someone used UTF-16 and characters outside the BMP?

On Windows I think `wchar_t` is 16bit and `mbstowcs()` is actually going to give you UTF-16, and it'll all work out OK.

On other platforms you're actually going to get values >= 0x10000 in `wpass[i]` and that's not going to work out so well.

I'm not sure I care.

Reply:
Weeellllll, have you seen what our PKCS#12 routines do with >0x10000 pass phrase characters? With regards to that, I think this does the right thing as the platform allows.